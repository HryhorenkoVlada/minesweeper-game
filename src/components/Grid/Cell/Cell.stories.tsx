import { Meta, StoryObj } from '@storybook/react'

import { Cell } from './Cell'

export default {
  title: 'Grid/Cell',
  component: Cell,
  argTypes: {
    coords: { defaultValue: [1, 1] },
  },
} as Meta

type Story = StoryObj<typeof Cell>

export const CellClosed: Story = {
  args: {
    children: 10,
  },
}
export const CellIsEmpty: Story = {
  args: {
    children: 0,
  },
}
export const CellWithBomb: Story = {
  args: {
    children: 9,
  },
}
export const CellWithFlag: Story = {
  args: {
    children: 11,
  },
}
export const CellWith1: Story = {
  args: {
    children: 1,
  },
}
export const CellWith2: Story = {
  args: {
    children: 2,
  },
}
export const CellWith5: Story = {
  args: {
    children: 5,
  },
}

export const CellWith8: Story = {
  args: {
    children: 8,
  },
}
