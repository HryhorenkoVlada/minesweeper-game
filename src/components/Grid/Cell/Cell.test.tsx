import React from 'react'
import { render, screen, fireEvent, createEvent } from '@testing-library/react'

import { CellState, Coords } from '@/types'

import { Cell, ClosedFrame } from './Cell'

describe('Cell component check', () => {
  const coords: Coords = [1, 1]
  const props = {
    coords,
    flagCounter: 0,
    bombs: 10,
    onClick: jest.fn(),
    onContextMenu: jest.fn(),
  }

  it('Empty cell renders correct', () => {
    const { asFragment } = render(<Cell {...props}>{CellState.empty}</Cell>)

    expect(asFragment()).toMatchSnapshot()
  })

  it('Bomb cell renders correct', () => {
    render(<Cell {...props}>{CellState.bomb}</Cell>)

    expect(screen.getByTestId(`bomb_${coords.join()}`)).toBeInTheDocument()
  })

  it('Flag cell renders correct', () => {
    render(<Cell {...props}>{CellState.flag}</Cell>)

    expect(screen.getByTestId(`flag_${coords.join()}`)).toBeInTheDocument()
  })

  it('Weak Flag cell renders correct', () => {
    render(<Cell {...props}>{CellState.weakFlag}</Cell>)

    expect(screen.getByTestId(`weakFlag_${coords.join()}`)).toBeInTheDocument()
  })

  it('Hidden cell renders correct', () => {
    const { asFragment } = render(<Cell {...props}>{CellState.hidden}</Cell>)

    expect(asFragment()).toMatchSnapshot()
  })

  it('Check prevent default contextMenu for empty type of cell', () => {
    render(<Cell {...props}>{CellState.empty}</Cell>)

    const cellComp = screen.getByTestId(`${coords}`)

    const contextMenuEvent = createEvent.contextMenu(cellComp)
    fireEvent(cellComp, contextMenuEvent)

    expect(contextMenuEvent.defaultPrevented).toBe(true)
  })

  it('onClick and onContextMenu handler should be called for active cells', () => {
    render(<Cell {...props}>{CellState.hidden}</Cell>)

    const cellComp = screen.getByTestId(`${coords}`)

    fireEvent.click(cellComp)
    fireEvent.contextMenu(cellComp)

    expect(props.onClick).toBeCalled()
    expect(props.onContextMenu).toBeCalled()
  })

  it('onClick and onContextMenu handler should not be called for active cells', () => {
    render(<Cell {...props}>{CellState.empty}</Cell>)

    const cellComp = screen.getByTestId(`${coords}`)

    fireEvent.click(cellComp)
    fireEvent.contextMenu(cellComp)

    expect(props.onClick).not.toBeCalled()
    expect(props.onContextMenu).not.toBeCalled()
  })

  it('Closed Frame renders correct', () => {
    const { asFragment } = render(<ClosedFrame mouseDown={true} />)
    expect(asFragment()).toMatchSnapshot()
  })
})
