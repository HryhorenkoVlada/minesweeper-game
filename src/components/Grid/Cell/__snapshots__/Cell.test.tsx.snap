// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cell component check Closed Frame renders correct 1`] = `
<DocumentFragment>
  <div
    class="css-8jxsla"
  />
</DocumentFragment>
`;

exports[`Cell component check Empty cell renders correct 1`] = `
<DocumentFragment>
  <div
    class="css-1l2qare"
    data-testid="1,1"
    role="cell"
  >
    0
  </div>
</DocumentFragment>
`;

exports[`Cell component check Hidden cell renders correct 1`] = `
<DocumentFragment>
  <div
    class="css-1tt7j02"
    data-testid="1,1"
    role="cell"
  >
    10
  </div>
</DocumentFragment>
`;
